### PR TITLE
HCAL LUT generation and validation 

### PR DESCRIPTION
**Pull request by akhukhun**
The PR includes 
- code to generate current HCAL LUTs: i.e. 2017 input/compression LUTs and HF channel mask, adc threshold and tdc mask tables. 
- Simple Hcal LUT analyzer for validation
- Increased precision for L1TriggerObject dumping that removes discrepancies due to rounding. 

The changes will not affect TP emulation.

### CODE REVIEW

File: CaloOnlineTools/HcalOnlineDb/src/WriteL1TriggerObjectsTxt.cc [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+{
+    using namespace edm;
+
+    edm::ESHandle<HcalDbService> conditions;
+    iSetup.get<HcalDbRecord>().get(conditions);
+
+    const HcalLutMetadata *metadata = conditions->getHcalLutMetadata();
+    const HcalTopology *topo = metadata->topo();
+
+    std::auto_ptr<HcalL1TriggerObjects> HcalL1TrigObjCol(new HcalL1TriggerObjects);
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/src/WriteL1TriggerObjectsTxt.cc]: auto_ptr is depricated in C++11.  Please use unique_ptr

File: CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc [hunk 2]
```diff
@@ -70,6 +70,7 @@ void CaloTPGTranscoderULUT::loadHCALCompress(HcalLutMetadata const& lutMetadata,
 	bool isHBHE = (abs(ieta) < theTrigTowerGeometry.firstHFTower(version)); 
 
         unsigned int lutsize = getOutputLUTSize(id);
+	outputLUT_[index].resize(lutsize);
 
         for (unsigned int i = 0; i < threshold; ++i)
```
Comment (rekovic) [inline CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc]: please use range-based loop.

File: CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc [hunk 3]
```diff
@@ -267,6 +268,8 @@ void HcaluLUTTPGCoder::update(const HcalDbService& conditions) {
      float gain = 0;
      uint32_t status = 0;
 
+     QIEType_[lutId]=conditions.getHcalQIEType(cell)->getValue();
+
      if (LUTGenerationMode_){
         const HcalCalibrations& calibrations = conditions.getHcalCalibrations(cell);
         for (int capId = 0; capId < 4; ++capId){
```
Comment (rekovic) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: use range-based loop

File: CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc [hunk 4]
```diff
@@ -409,6 +412,16 @@ std::vector<unsigned short> HcaluLUTTPGCoder::getLinearizationLUTWithMSB(const H
...
+    int type=QIEType_[lutId];
+    const std::vector<Lut>& lut= type==0 ? inputLUT_ : 
+				 type==1 ? upgradeQIE10LUT_ : 
+					   upgradeQIE11LUT_ ;
+    return lut.at(lutId);
+}
+
 void HcaluLUTTPGCoder::lookupMSB(const HBHEDataFrame& df, std::vector<bool>& msb) const{
   msb.resize(df.size());
   for (int i=0; i<df.size(); ++i)
```
Comment (rekovic) [inline CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc]: use range-based loop.

File: CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    xmls1.create_lut_map();
+    xmls2.create_lut_map();
+
+    const char* DET[7]={"EMPTY", "HB" , "HE", "HO", "HF", "HT", "OTHER"};
+    int ntota[7]; 
+    int nzero[7]; 
+    int nreal[7]; 
+    int extra[7];
+    for(int i=0; i<7; ++i) {
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc]: use range-based loop

File: CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	HcalGenericDetId::HcalGenericSubdetector subdet = id.genericSubdet();
+	ntota[subdet]++;
+	if(xml1==xmls1.end()){
+	    extra[subdet]++;
+	    cout << "FAIL: DetId " << id << "(" << id.rawId() <<") IS PRESENT IN " << file2 << "  BUT ABSENT IN " <<  file1  << " " << id << endl;
+	    //return;
+	}
+	bool zero=true;
+	const std::vector<unsigned int>& lut2 = xml2->second;
+	for(size_t i=0; i<lut2.size(); ++i){
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc]: use range-based loop

File: CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	    if(lut2[i]>0){
+		zero=false;
+	    }
+	}
+	if(zero) nzero[subdet]++;
+	else     nreal[subdet]++;
+    }
+    cout << Form("%3s:  %8s  %8s  %8s  %8s", "Det", "total", "nonzeros", "zeroes", "extra") << endl; 
+    for(int i=1; i<6; ++i) cout << Form("%3s:  %8d  %8d  %8d  %8d", DET[i], ntota[i], nreal[i], nzero[i], extra[i]) << endl;
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc]: use range-based loop

File: CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    cout << "--------------------------------------------" << endl;
+    for(int i=0; i<7; ++i) {
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc]: use range-based loop

Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/bin/hcalLUT.cc]: and everywhere else in this file.

File: CaloOnlineTools/HcalOnlineDb/src/HcalLutManager.cc [hunk 9]
```diff
@@ -1550,6 +1643,11 @@ int HcalLutManager::createLutXmlFiles_HBEFFromCoder_HOFromAscii_ZDC( std::string
   //
   const std::map<int, boost::shared_ptr<LutXml> > _comp_lut_xml = getCompressionLutXmlFromCoder( _transcoder, _tag, split_by_crate );
   addLutMap( xml, _comp_lut_xml );
+
+  for(int masktype=0; masktype<3; ++masktype){
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/src/HcalLutManager.cc]: use range-based loop

File: CaloOnlineTools/HcalOnlineDb/src/WriteL1TriggerObjectsTxt.cc [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	HcalDetId cell(id);
+	HcalSubdetector subdet = cell.subdet();
+	if (subdet != HcalBarrel and subdet != HcalEndcap and subdet != HcalForward) continue;
+
+        HcalCalibrations calibrations = conditions->getHcalCalibrations(cell);
+
+	float gain = 0.0;
+	float ped = 0.0;
+       
+	for (int i=0; i<4; ++i) {
```
Comment (rekovic) [inline CaloOnlineTools/HcalOnlineDb/src/WriteL1TriggerObjectsTxt.cc]: use range-based loop

**Comment by franzoni** [comment]
please test

**Comment by perrotta** [comment]
-1
wrong rebase

**Comment by kpedro88** [comment]
-1

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by arunhep** [comment]
+1
changes in HLT directory are because of this extra commit https://github.com/cms-sw/cmssw/commit/8e59646c5dc2756e96963b11dceef23e96c92c1b

**Comment by ggovi** [comment]
+1

File: CaloOnlineTools/HcalOnlineDb/src/HcalLutAnalyzer.cc [hunk 11]
```diff
@@ -0,0 +1,541 @@
...
+
+
+void
+HcalLutAnalyzer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
+}
+
+DEFINE_FWK_MODULE(HcalLutAnalyzer);
```
Comment (davidlange6) [inline CaloOnlineTools/HcalOnlineDb/src/HcalLutAnalyzer.cc]: Hi @akhukhun - CaloOnlineTools/HcalOnlineDb/src can either be a library or define plugins, but not both. You should probably move this class to the plugins directory.

**Comment by kpedro88** [comment]
please test

**Comment by arunhep** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Comment by rekovic** [comment]
+1

**Pull request merged**